compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCastGen.java

```diff
@@ -310,122 +379,174 @@ private static void generateJToBCheckCast(MethodVisitor mv, BIRVarToJVMIndexMap
-                //} else if (targetType is bir:BFiniteType) {
-                //    generateCheckCastToFiniteType(mv, sourceType, targetType);
-                //    return;
-                //} else if (sourceType is bir:BRecordType && (targetType is bir:BMapType && targetType.constraint
-                // is bir:BTypeAny)) {
-                //    // do nothing
-            }
+            return;
+        }
 
-            checkCast(mv, targetType);
-            String targetTypeClass = getTargetClass(targetType);
-            if (targetTypeClass != null) {
-                mv.visitTypeInsn(CHECKCAST, targetTypeClass);
-            }
+        switch (targetType.tag) {
+            case TypeTags.FLOAT:
+                generateCheckCastJToBFloat(mv, sourceType);
+                break;
+            case TypeTags.DECIMAL:
+                generateCheckCastJToBDecimal(mv, sourceType);
+                break;
+            case TypeTags.BOOLEAN:
+                generateCheckCastJToBBoolean(mv, sourceType);
+                break;
+            case TypeTags.BYTE:
+                generateCheckCastJToBByte(mv, sourceType);
+                break;
+            case TypeTags.NIL:
+            case TypeTags.NEVER:
+                // Do nothing
+                break;
+            default:
+                switch (targetType.tag) {
+                    case TypeTags.UNION:
+                        generateCheckCastJToBUnionType(mv, indexMap, sourceType, (BUnionType) targetType);
+                        break;
+                    case TypeTags.ANYDATA:
+                        generateCheckCastJToBAnyData(mv, indexMap, sourceType);
+                        break;
+                    case TypeTags.HANDLE:
+                        generateJCastToBHandle(mv, sourceType);
+                        break;
+                    case TypeTags.ANY:
+                        generateJCastToBAny(mv, indexMap, sourceType, targetType);
+                        break;
+                    case TypeTags.JSON:
+                        generateCheckCastJToBJSON(mv, indexMap, sourceType);
+                        break;
+                    case TypeTags.FINITE:
+                        generateCheckCastJToBFiniteType(mv, indexMap, sourceType, targetType);
+                        // TODO fix below properly - rajith
+                        //} else if (sourceType is bir:BXMLType && targetType is bir:BMapType) {
+                        //    generateXMLToAttributesMap(mv, sourceType);
+                        //    return;
+                        //} else if (targetType is bir:BFiniteType) {
+                        //    generateCheckCastToFiniteType(mv, sourceType, targetType);
+                        //    return;
+                        //} else if (sourceType is bir:BRecordType && (targetType is bir:BMapType &&
+                        // targetType.constraint
+                        // is bir:BTypeAny)) {
+                        //    // do nothing
```

Comment:
Shall we remove the commented code? We can add a Todo if needed about what to fix properly and remove the code. Anyone can see it from git history right?